addons/barcodes_gs1_nomenclature/static/src/js/barcode_parser.js

```
@@ -0,0 +1,151 @@
+            }
+        } else if (rule.gs1_content_type === 'identifier'){
+            if (parseInt(match[2][match[2].length - 1]) !== this.get_barcode_check_digit("0".repeat(18 - match[2].length) + match[2])){
+                return null; // Warning ?
+            }
+            result.value = match[2];
+        } else if (rule.gs1_content_type === 'date'){
+            if (match[2].length !== 6){
+                return null; // Warning ?
+            }
+            result.value = this.gs1_date_to_date(match[2]);
+        } else {
+            result.value = match[2];
+        }
+        return result;
+    },
+
+    /**
+     * Try to decompose the gs1 extanded barcode into several unit of information using gs1 rules.
+     *
+     * @param {string} barcode
+     * @returns {Array} Array of object
+     */
+    gs1_decompose_extanded: function(barcode) {
+        const results = [];
+        const rules = this.nomenclature.rules.filter(rule => rule.encoding === 'gs1-128');
+        let separatorReg = FNC1_CHAR + "?";
+        if (this.nomenclature.gs1_separator_fnc1 && this.nomenclature.gs1_separator_fnc1.trim()){
+            separatorReg = `(?:${this.nomenclature.gs1_separator_fnc1})?`;
+        }
+
+        while (barcode.length > 0) {
+            const barcodeLength = barcode.length;
+            for (const rule of rules) {
+                const match = barcode.match("^" + rule.pattern + separatorReg);
+                if (match && match.length >= 3) {
+                    const res = this.parse_gs1_rule_pattern(match, rule);
+                    if (res) {
+                        barcode = barcode.slice(match.index + match[0].length);
+                        results.push(res);
+                        if (barcode.length === 0) {
+                            return results; // Barcode completly parsed, no need to keep looping.
+                        }
+                    } else {
+                        return null;
```

maybe should return an exception if we could not parse the barcode